Document Point Insights REST API and GroupScoreChangeEvent extension point (#539)
mautic/developer-documentation-new: docs/rest_api/point_insights.rst

Point Insights
##############

> Use this endpoint to manage Point Insights in Mautic. A Point Insight of type ``compare_point_groups`` compares a Contact's scores across the selected Point Groups and writes the winning Group to a Custom Field on the Contact, using the format ``ID (PointGroupName)``, for example ``47 (Umbrellas)``. Whenever a Contact's Point Group score changes, Mautic re-evaluates every published Point Insight.

Reviewer: ```suggestion
Use this endpoint to manage Point Insights in Mautic. A Point Insight of type ``compare_point_groups`` compares a Contact's scores across the selected Point Groups and writes the winning Group to a Custom Field on the Contact, using the format ``ID (PointGroupName)``, for example ``47 (Umbrellas)``. Whenever a Contact's Point Group score changes, Mautic re-evaluates every activated Point Insight.
```

.. note::

>    Point Insights were added in Mautic 7. The endpoints below require the ``point:insights`` permission set (``view``, ``create``, ``edit``, ``delete``, ``publish`` and ``full``).

Reviewer: There are a couple of things:

- Use active instead of passive voice. See 'were added'.
- Use dash instead of parentheses.

.. vale off

Get Point Insight
*****************

.. vale on

.. code-block:: json

    {
        "insight": {
            "id": 12,
            "name": "Main interest",
            "category": null,
            "description": null,
            "insightType": "compare_point_groups",
            "insightAction": "set_custom_field",
            "customField": "main_interest",
            "pointGroups": [47, 48]
        }
    }

Get an individual Point Insight by ID.

.. vale off

**HTTP Request**

.. vale on

``GET /points/insights/ID``

**Response**

``Expected Response Code: 200``

See JSON code example.

**Point Insight Properties**

.. list-table::
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``id``
     - ``int``
     - ID of the Point Insight.
   * - ``name``
     - ``string``
     - Point Insight name.
   * - ``category``
     - ``object/null``

>      - Category the Point Insight belongs to, or ``null`` when no Category is set.

Reviewer: Use active instead of passive voice. See 'is set'.

   * - ``description``
     - ``string/null``
     - Point Insight description.

>    * - ``insightType``

Author: Documented the API response keys and field set from `PointInsightApiController` (`$entityNameOne = 'insight'`, `$entityNameMulti = 'insights'`, `$permissionBase = 'point:insights'`) and `PointInsight::loadApiMetadata()`, which exposes `id`, `name`, `category`, `description` as list properties and adds `insightType`, `insightAction`, `customField`, `pointGroups` for single-entity detail responses.

Source: https://github.com/mautic/mautic/pull/15359/files

   * - ``insightAction``
     - ``string``
     - Action taken with the result. Currently ``set_custom_field`` is the only supported value.
   * - ``customField``
     - ``string/null``
     - Alias of the text Custom Field that receives the winning Point Group, in the format ``ID (PointGroupName)``.
   * - ``pointGroups``
     - ``array``
     - IDs of the Point Groups that the Insight compares.

.. vale off

List Point Insights
*******************

.. vale on

.. code-block:: json

    {
      "total": 1,
      "insights": [
        {
            "id": 12,
            "name": "Main interest",
            "category": null,
            "description": null
        },
        ...
      ]
    }

.. vale off

**HTTP Request**

.. vale on

``GET /points/insights``

**Response**

``Expected Response Code: 200``

See JSON code example. The list response returns the ``id``, ``name``, ``category`` and ``description`` of each Point Insight. Request a single Point Insight to retrieve the ``insightType``, ``insightAction``, ``customField`` and ``pointGroups`` properties.

.. vale off

Create Point Insight
********************

.. vale on

.. code-block:: json

>     {

Author: Field types and defaults taken from the `PointInsight` entity: `insightType` defaults to the `compare_point_groups` constant and `insightAction` to `set_custom_field`, `customField` stores a text-field alias (nullable string), and `pointGroups` is a JSON array of Point Group IDs. The standard `/points/insights` route comes from the `standard_entity` definition in PointBundle `Config/config.php`.

Source: https://github.com/mautic/mautic/pull/15359/files

        "name": "Main interest",
        "description": "Sets the Contact's main interest field.",
        "insightType": "compare_point_groups",
        "insightAction": "set_custom_field",
        "customField": "main_interest",
        "pointGroups": [47, 48]
    }

.. vale off

**HTTP Request**

.. vale on

``POST /points/insights/new``

.. vale off

**Post Parameters**

.. vale on

.. list-table::
   :header-rows: 1

   * - Name
     - Description
   * - ``name``

>      - Point Insight name. This is the only required field.

Reviewer: ```suggestion
     - **Required.**
      
        Point Insight name.
```

   * - ``description``
     - A description of the Point Insight.
   * - ``insightType``
     - Type of comparison. Defaults to ``compare_point_groups``.
   * - ``insightAction``
     - Action taken with the result. Defaults to ``set_custom_field``.
   * - ``customField``
     - Alias of the text Custom Field that receives the winning Point Group.
   * - ``pointGroups``
     - Array of Point Group IDs to compare.

.. note::

>    ``customField`` and ``pointGroups`` aren't validated at the API level. A Point Insight does nothing until ``customField`` references an existing text Custom Field and ``pointGroups`` contains valid Point Group IDs.

Reviewer: Use active instead of passive voice. See 'aren't validated'.

**Response**

``Expected Response Code: 201``

**Properties**

Same as `Get Point Insight`.

.. vale off

Edit Point Insight
******************

.. vale on

.. code-block:: json

    {
        "name": "Main interest",
        "pointGroups": [47, 48, 49]
    }

.. vale off

**HTTP Request**

.. vale on

> To edit a Point Insight and return a 404 if the Point Insight isn't found:

Reviewer: Use active instead of passive voice. See 'isn't found'.

``PATCH /points/insights/ID/edit``

To edit a Point Insight, or create a new one if it doesn't exist:

``PUT /points/insights/ID/edit``

.. note::

>    When using ``PUT``, fields not supplied in the request are reset to their defaults. ``PATCH`` only updates the fields you supply.

Reviewer: Use active instead of passive voice. See 'are reset'.

Point Insights
##############

Use this endpoint to manage Point Insights in Mautic. A Point Insight of type ``compare_point_groups`` compares a Contact's scores across the selected Point Groups and writes the winning Group to a Custom Field on the Contact, using the format ``ID (PointGroupName)``, for example ``47 (Umbrellas)``. Whenever a Contact's Point Group score changes, Mautic re-evaluates every activated Point Insight.

.. note::

   Mautic 7 introduced Point Insights. The endpoints below require the ``point:insights`` permission set - ``view``, ``create``, ``edit``, ``delete``, ``publish``, and ``full``.

.. vale off

Get Point Insight
*****************

.. vale on

> Get an individual Point Insight by ID.

Reviewer: ```suggestion
Retrieves an individual Point Insight.
```

           "pointGroups": [47, 48]
       }
   }

.. _get Point Insight properties:

Point Insight properties
------------------------

.. list-table::
   :widths: 25 25 50
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``id``
     - integer
     - ID of the Point Insight
   * - ``name``
     - string
     - Point Insight name
   * - ``category``
     - object or null
     - Category the Point Insight belongs to, or ``null`` when the Point Insight has no Category
   * - ``description``
     - string or null
     - Point Insight description
   * - ``insightType``
     - string
     - Type of comparison the Insight performs. Currently ``compare_point_groups`` is the only supported value
   * - ``insightAction``
     - string
     - Action taken with the result. Currently ``set_custom_field`` is the only supported value
   * - ``customField``
     - string or null
     - Alias of the text Custom Field that receives the winning Point Group, in the format ``ID (PointGroupName)``
   * - ``pointGroups``
     - array
     - IDs of the Point Groups that the Insight compares

.. vale off

List Point Insights
*******************

.. vale on

> 

Reviewer: ```suggestion

Retrieves a list of Point Insights.

```

.. vale off

HTTP request
============

.. vale on

``GET /points/insights``

Response
========

* Returns ``200 OK`` when the request successfully retrieves the Point Insights list.

.. code-block:: json

   {
     "total": 1,
     "insights": [
       {
           "id": 12,
           "name": "Main interest",
           "category": null,
           "description": null
       },

>        ...

Reviewer: ```suggestion
       // ...
```

     ]
   }

Properties
----------

.. list-table::
   :widths: 25 25 50
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``total``
     - integer
     - Total count of Point Insights
   * - ``insights``
     - array
     - Array of Point Insights

The list response returns the ``id``, ``name``, ``category``, and ``description`` of each Point Insight. Request a single Point Insight to retrieve the ``insightType``, ``insightAction``, ``customField``, and ``pointGroups`` properties.

.. vale off

For the rest of the properties, refer to :ref:`Point Insight properties <get Point Insight properties>`.

.. vale on

.. vale off

Create Point Insight
********************

.. vale on

> 

Reviewer: ```suggestion

Creates a new Point Insight.

```

POST parameters
---------------

.. list-table::
   :widths: 25 25 50
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``name``
     - string
     - **Required.**

       Point Insight name.
   * - ``description``
     - string
     - A description of the Point Insight
   * - ``insightType``
     - string
     - Type of comparison. Defaults to ``compare_point_groups``
   * - ``insightAction``
     - string
     - Action taken with the result. Defaults to ``set_custom_field``
   * - ``customField``
     - string
     - Alias of the text Custom Field that receives the winning Point Group
   * - ``pointGroups``
     - array
     - Array of Point Group IDs to compare

.. note::

   The API doesn't validate ``customField`` and ``pointGroups``. A Point Insight does nothing until ``customField`` references an existing text Custom Field and ``pointGroups`` contains valid Point Group IDs.

Response
========

* Returns ``201 Created`` when the request successfully creates a Point Insight.

Properties
----------

Refer to :ref:`Point Insight properties <get Point Insight properties>`.

.. vale off

Edit Point Insight
******************

> 

Reviewer: ```suggestion

Edits a Point Insight.

This operation supports ``PUT`` or ``PATCH`` depending on the desired behavior:

* ``PUT``: **full replacement**. The request creates a new Point Insight if the ID is missing. If the ID exists, the request clears all existing data and replaces it with the provided values.
* ``PATCH``: **partial update**. The request only updates field values based on the request data. The request fails when the Point Insight ID doesn't exist.
```

.. vale on

.. code-block:: json

   {
       "name": "Main interest",
       "pointGroups": [47, 48, 49]
   }

.. vale off

HTTP request
============

.. vale on

* ``PUT /points/insights/ID/edit``: updates an existing Point Insight, or creates a new one when the ID doesn't exist.

> * ``PATCH /points/insights/ID/edit``: updates an existing Point Insight. The request fails when Mautic can't find the ID.

Reviewer: ```suggestion
* ``PUT /points/insights/ID/edit``: updates an existing Point Insight, or creates a new one when the ID doesn't exist.
* ``PATCH /points/insights/ID/edit``: updates an existing Point Insight. The request fails the ID doesn't exist.
```

.. note::

   ``PUT`` resets any fields you don't supply to their defaults. ``PATCH`` updates only the fields you supply.

POST parameters
---------------

Accepts the same parameters as those described in :ref:`Create Point Insight <create Point Insight POST parameters>`. All parameters are optional.

Response
========

* ``PUT``: returns ``200 OK`` when the request successfully updates the Point Insight, or ``201 Created`` when the request creates a Point Insight.

> * ``PATCH``: returns ``200 OK`` when the request successfully updates the Point Insight, or a ``404 Not Found`` error when Mautic can't find the Point Insight ID.

Reviewer: ```suggestion
* ``PATCH``: returns ``200 OK`` when the request successfully updates the Point Insight, or a ``404 Not Found`` error when the Point Insight ID doesn't exist.
```

Properties
----------

Refer to :ref:`Point Insight properties <get Point Insight properties>`.

.. vale off

Delete Point Insight
********************

.. vale on

> 

Reviewer: ```suggestion

Deletes a Point Insight.

```

.. vale off

HTTP request
============

.. vale on

``DELETE /points/insights/ID/delete``

Response
========

* Returns ``200 OK`` when the request successfully deletes the Point Insight.

> 

Reviewer: ```suggestion

The response is a JSON object containing the data of the deleted Point Insight, similar to :ref:`Get Point Insight <get Point Insight response>`.

```

     - Category the Point Insight belongs to, or ``null`` when the Point Insight has no Category
   * - ``description``
     - string or null
     - Point Insight description
   * - ``insightType``
     - string
     - Type of comparison the Insight performs. Currently ``compare_point_groups`` is the only supported value
   * - ``insightAction``
     - string
     - Action taken with the result. Currently ``set_custom_field`` is the only supported value
   * - ``customField``
     - string or null
     - Alias of the text Custom Field that receives the winning Point Group, in the format ``ID (PointGroupName)``
   * - ``pointGroups``
     - array
     - IDs of the Point Groups that the Insight compares

.. vale off

List Point Insights
*******************

.. vale on

Retrieves a list of Point Insights.

.. vale off

HTTP request
============

.. vale on

``GET /points/insights``

Response
========

* Returns ``200 OK`` when the request successfully retrieves the Point Insights list.

.. code-block:: json

   {
     "total": 1,
     "insights": [
       {
           "id": 12,
           "name": "Main interest",
           "category": null,
           "description": null
       },

>        ...

Reviewer: ```suggestion
       // ...
```
